apps/api/src/database/repositories/balance-repository.ts

```diff
@@ -246,6 +198,136 @@ async function resetAgentBalancesImpl(
   }
 }
 
+/**
+ * Atomic balance update within a transaction
+ * @param tx Database transaction
+ * @param agentId Agent ID
+ * @param tokenAddress Token address
+ * @param amountDelta Positive for increment, negative for decrement
+ * @param specificChain Specific chain for the token
+ * @param symbol Token symbol
+ * @returns The new balance amount
+ */
+async function updateBalanceInTransactionImpl(
+  tx: DatabaseTransaction,
+  agentId: string,
+  tokenAddress: string,
+  amountDelta: number,
+  specificChain: SpecificChain,
+  symbol: string,
+): Promise<number> {
+  const [result] = await tx
+    .update(balances)
+    .set({
+      amount: sql`${balances.amount} + ${amountDelta}`,
+      updatedAt: new Date(),
+      specificChain,
+      symbol,
+    })
+    .where(
+      and(
+        eq(balances.agentId, agentId),
+        eq(balances.tokenAddress, tokenAddress),
+      ),
+    )
+    .returning();
+
+  if (!result) {
+    if (amountDelta > 0) {
+      // Create balance if it doesn't exist for increments
+      const [newResult] = await tx
+        .insert(balances)
+        .values({
+          agentId,
+          tokenAddress,
+          amount: amountDelta,
+          specificChain,
+          symbol,
+          createdAt: new Date(),
+          updatedAt: new Date(),
+        })
+        .returning();
+
+      if (!newResult) {
+        throw new Error("Failed to create new balance");
+      }
+
+      return newResult.amount;
+    } else {
+      // For decrements, the balance should already exist
+      throw new Error(
+        `Balance not found for agent ${agentId}, token ${tokenAddress}`,
+      );
+    }
+  } else if (amountDelta < 0 && result.amount < 0) {
+    throw new Error(
```

Comment:
This works well afaict, but curious if there's any advantage to using `tx.rollback()` instead of throwing?

Reply:
I guess I assumed that drizzle would roll back the transaction automatically if an Error is thrown during transaction execution

Reply:
Calling `rollback()` is literally just implemented as throwing an Error: https://github.com/drizzle-team/drizzle-orm/blob/8b8d78ec3d209ce3224b609d633b7ce090092889/drizzle-orm/src/pg-core/session.ts#L257